pkg/vsphere/toolbox/process.go

```diff
@@ -0,0 +1,359 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package toolbox
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+)
+
+var (
+	xmlEscape *strings.Replacer
+
+	shell = "/bin/sh"
+
+	defaultOwner = os.Getenv("USER")
+)
+
+func init() {
+	// See: VixToolsEscapeXMLString
+	chars := []string{
+		`"`,
+		"%",
+		"&",
+		"'",
+		"<",
+		">",
+	}
+
+	replace := make([]string, 0, len(chars)*2)
+
+	for _, c := range chars {
+		replace = append(replace, c)
+		replace = append(replace, url.QueryEscape(c))
+	}
+
+	xmlEscape = strings.NewReplacer(replace...)
+
+	// See procMgrPosix.c:ProcMgrStartProcess:
+	// Prefer bash -c as is uses exec() to replace itself,
+	// whereas bourne shell does a fork & exec, so two processes are started.
+	if sh, err := exec.LookPath("bash"); err != nil {
+		shell = sh
+	}
+
+	if defaultOwner == "" {
+		defaultOwner = "toolbox"
+	}
+}
+
+// ProcessState is the toolbox representation of the GuestProcessInfo type
+type ProcessState struct {
+	Name      string
+	Args      string
+	Owner     string
+	Pid       int64
+	ExitCode  int32
+	StartTime int64
+	EndTime   int64
+}
+
+func (s *ProcessState) toXML() string {
+	const format = "<proc>" +
+		"<cmd>%s</cmd>" +
+		"<name>%s</name>" +
+		"<pid>%d</pid>" +
+		"<user>%s</user>" +
+		"<start>%d</start>" +
+		"<eCode>%d</eCode>" +
+		"<eTime>%d</eTime>" +
+		"</proc>"
+
+	name := filepath.Base(s.Name)
+
+	argv := []string{s.Name}
+
+	if len(s.Args) != 0 {
+		argv = append(argv, xmlEscape.Replace(s.Args))
+	}
+
+	args := strings.Join(argv, " ")
+
+	exit := atomic.LoadInt32(&s.ExitCode)
+	end := atomic.LoadInt64(&s.EndTime)
+
+	return fmt.Sprintf(format, name, args, s.Pid, s.Owner, s.StartTime, exit, end)
+}
+
+// Process managed by the ProcessManager.
+type Process struct {
+	ProcessState
+
+	Start func(context.Context, *VixMsgStartProgramRequest) (int64, error)
+	Wait  func() error
+	Kill  context.CancelFunc
+
+	ctx context.Context
+}
+
+// ProcessError can be returned by the Process.Wait function to propagate ExitCode to ProcessState.
+type ProcessError struct {
+	Err      error
+	ExitCode int32
+}
+
+func (e *ProcessError) Error() string {
+	return e.Err.Error()
+}
+
+// ProcessManager manages processes within the guest.
+// See: http://pubs.vmware.com/vsphere-60/topic/com.vmware.wssdk.apiref.doc/vim.vm.guest.ProcessManager.html
+type ProcessManager struct {
+	wg      sync.WaitGroup
+	mu      sync.Mutex
+	expire  time.Duration
+	entries map[int64]*Process
+	pids    sync.Pool
+}
+
+// NewProcessManager creates a new ProcessManager instance.
+func NewProcessManager() *ProcessManager {
+	pid := int64(32768) // TODO: /proc/sys/kernel/pid_max
+
+	return &ProcessManager{
+		expire:  time.Minute * 5,
+		entries: make(map[int64]*Process),
+		pids: sync.Pool{
+			New: func() interface{} {
+				return atomic.AddInt64(&pid, 1)
```

Comment:
I think I'm confused a bit as it is not clear to me why we are using sync.Pool for pids instead of []int64? Hence maybe a comment would be useful here :)